iffdigest.cc

```diff
@@ -53,6 +54,56 @@ IFFChunk::operator=(const IFFChunk& ck)
   }
 }
 
+void
+IFFChunk::writeData(unsigned int &len, char* outData, const enum IFFFormat iffFormat)
+{
+  unsigned int listLenStart = len;
+  if(ctype == IFF_CHUNK_LIST) {
+    // 'LIST'
+    if(outData) {
+      memcpy(outData+len, "LIST", 4);
+    }
+    len += 4;
+    listLenStart = len;
+    // LISTlength unknown here -> set below
+    len += 4;
+  }
+
+  // chunk ID
+  if(outData) {
+    memcpy(outData+len, &ckid, sizeof(iff_ckid_t));
+  }
+  len+=sizeof(iff_ckid_t);
+
+  if(ctype == IFF_CHUNK_DATA) {
+    // chunk len
+    if(outData) {
+      unsigned int chunkLen = u32(length, iffFormat);
+      memcpy(outData+len, &chunkLen, sizeof(chunkLen));
+    }
+    len+=4;
+
+    // chunk data
+    if(outData) {
+      memcpy(outData+len, data, length);
+    }
+    len += length;
+  }
+
+  // Write all sub chunks recursively
+  IFFChunkIterator subChunkIterator;
+  for(subChunkIterator = ck_begin();
+      subChunkIterator != ck_end(); subChunkIterator++) {
+    (*subChunkIterator).writeData(len, outData, iffFormat);
+  }
+  // ListLen
+  if(ctype == IFF_CHUNK_LIST && outData) {
+    // length itself is not part of length -> '- sizeof(unsigned int)'
+    unsigned int listLen = u32(len - listLenStart - sizeof(unsigned int), iffFormat);
+    memcpy(outData+listLenStart, &listLen, sizeof(listLen));
+  }
+}
```

Comment:
Some bounds checking could be good; perhaps with a `maxLen` parameter that the function stops writing if `len` ever exceeds this.
(Also, should the type of an offset be `int` or something like `size_t`?)